xt_imager.py

```diff
@@ -0,0 +1,259 @@
+        image_size = 0
+    else:
+        f_img = open(args.image, "rb")
+
+    bytes_sent = 0
+    block_start = base_addr // mmc_block_size
+    out_fullname = os.path.join(tftp_root, chunk_filename)
+
+    if args.serverip:
+        conn_send(conn, f"env set serverip {args.serverip}\r")
+        conn_wait_for(conn, uboot_propmt)
+
+    if args.ipaddr:
+        conn_send(conn, f"env set ipaddr {args.ipaddr}\r")
+        conn_wait_for(conn, uboot_propmt)
+
+    # switch to the required MMC device/partition
+    conn_send(conn, f"mmc dev {mmc_device} {mmc_part}\r")
+    conn_wait_for(conn, uboot_propmt)
+
+    # do in loop:
+    # - read X MB chunk from image file
+    # - save chunk to file in tftp root
+    # - tell u-boot to 'tftp-and-emmc' chunk
+    while True:
+        if use_lzma:
+            data = f_lzma.read(chunk_size_in_bytes)
+        else:
+            data = f_img.read(chunk_size_in_bytes)
+
+        if not data:
+            break
+
+        chunk_size_in_blocks = len(data) // mmc_block_size
+        if len(data) % mmc_block_size:
+            chunk_size_in_blocks += 1
+
+        buffer_is_00_only = True
+        for i in range(len(data)):
+            # check for zero only
+            if (data[i] != 0):
+                buffer_is_00_only = False
+
+        if buffer_is_00_only:
+            conn_send(conn, f"mw.b 0x48000000 0x00 0x{len(data):X}\r")
+            conn_wait_for(conn, uboot_propmt)
+        else:
+            # create chunk
+            f_out = open(out_fullname, "wb")
+            f_out.write(data)
+            f_out.close()
+
+            conn_send(conn, f"tftp 0x48000000 {chunk_filename}\r")
+            # check that all bytes are transmitted
+            conn_wait_for(conn, f"Bytes transferred = {len(data)}")
+            conn_wait_for(conn, uboot_propmt)
+
+        conn_send(conn, f"mmc write 0x48000000 0x{block_start:X} 0x{chunk_size_in_blocks:X}\r")
+        # check that all blocks are written properly
+        conn_wait_for(conn, f"{chunk_size_in_blocks} blocks written: OK")
+        conn_wait_for(conn, uboot_propmt)
+
+        bytes_sent += len(data)
+        block_start += chunk_size_in_blocks
+
+        if image_size:
+            print(f"\nProgress: {bytes_sent:_}/{image_size:_} ({bytes_sent * 100 // image_size}%)")
+        else:
+            print(f"\nProgress: {bytes_sent:_}")
+
+        print("===============================")
+
+    # send "newline char" to start further output on the new line
+    print("")
+
+    os.remove(out_fullname)
+    if use_lzma:
+        f_lzma.close()
+    else:
+        f_img.close()
+    conn.close()
+
+    log.info("Image was flashed successfully.")
+
+
+def open_connection(args):
+    # Default value
+    dev_name = '/dev/ttyUSB0'
+    if args.serial:
+        dev_name = args.serial
+    baud = 115200
+
+    log.info(f"Using serial port {dev_name} with baudrate {baud}")
+    conn = serial.Serial(port=dev_name, baudrate=baud, timeout=20)
```

Comment:
maybe I am wrong, but call of Serial(port=dev_name, baudrate=baud, timeout=20) opens the port, after that we close it and open again
i think, it is not necessary , it is enough to call Serial(port=dev_name, baudrate=baud, timeout=20) and that is it

Reply:
You are right! Thanks! I have cleaned this code.